distribution/integration-tests/src/test/java/org/jboss/tools/rsp/itests/wildfly/WildFlyPublishingTest.java

```diff
@@ -0,0 +1,172 @@
+        serverProxy.publish(pubReq);
+        state = waitForNonNullServerState(5);
+        assertNotNull(state);
+        list = state.getDeployableStates();
+        assertNotNull(list);
+        assertEquals(1, list.size());
+        ds1 = list.get(0);
+        assertNotNull(ds1);
+        assertEquals(ServerManagementAPIConstants.PUBLISH_STATE_NONE, ds1.getPublishState());
+        assertEquals(ServerManagementAPIConstants.STATE_STARTED, ds1.getState());
+        
+        // TODO touch the file
+        war.setLastModified(System.currentTimeMillis());
+        state = waitForNonNullServerState(5);
+        assertNotNull(state);
+        list = state.getDeployableStates();
+        assertNotNull(list);
+        assertEquals(1, list.size());
+        ds1 = list.get(0);
+        assertNotNull(ds1);
+        assertEquals(ServerManagementAPIConstants.PUBLISH_STATE_INCREMENTAL, ds1.getPublishState());
+        assertEquals(ServerManagementAPIConstants.STATE_STARTED, ds1.getState());
+
+        
+        ModifyDeployableRequest req2 = new ModifyDeployableRequest(handle, deployable);
+        serverProxy.removeDeployable(req2);
+        state = waitForNonNullServerState(5);
+        assertNotNull(state);
+        list = state.getDeployableStates();
+        assertNotNull(list);
+        assertEquals(1, list.size());
+        ds1 = list.get(0);
+        assertNotNull(ds1);
+        assertEquals(ServerManagementAPIConstants.PUBLISH_STATE_REMOVE, ds1.getPublishState());
+        assertEquals(ServerManagementAPIConstants.STATE_STARTED, ds1.getState());
+
+        // TODO publish
+        pubReq = new PublishServerRequest(handle, ServerManagementAPIConstants.PUBLISH_FULL);
+        serverProxy.publish(pubReq);
+        state = waitForNonNullServerState(5);
+        assertNotNull(state);
+        list = state.getDeployableStates();
+        assertNotNull(list);
+        assertEquals(0, list.size());
+                
+        serverProxy.stopServerAsync(new StopServerAttributes("wildfly6a", true)).get();
+        waitForServerState("stopped", 10);
+    }
+
+	protected File createWar() {
+		Path deployments = null;
+		File war = null;
+		try {
+			deployments = Files.createTempDirectory(getClass().getName() + "5");
+			war = deployments.resolve(WAR_FILENAME).toFile();
+			if (!(new DeploymentGeneration().createWar(war))) {
+				fail();
+			}
+		} catch (IOException e) {}
+		return war != null && war.exists() && war.isFile() ? war : null;
+	}
+	
+    private void waitForServerState(String serverState, int timeout) throws Exception {
+        int tries = timeout;
+        
+        while(tries > 0) {
+            tries--;
+            String s = client.getStateString();
+            System.out.println("State String: " + s);
+            if (serverState.equals(s)) {
+            	client.clearState();
+                return;
+            }
+            Thread.sleep(1000);
+        }
```

Comment:
l145-154 could be replaced by a much simpler timely limited poll from a ```ArrayBlockingQueue``` (that gets filled up by the listener) in the TestClient:
I did exactly that in ```PollThreadTest```:
https://github.com/redhat-developer/rsp-server/blob/master/framework/tests/org.jboss.tools.rsp.server.spi.test/src/main/java/org/jboss/tools/rsp/server/spi/model/polling/PollThreadTest.java#L336